File: python/sorting/quick_sort.py
import time
from utils.arrays import create_random_array


def partition(array, low, high):
    moving_target = low + 1
    index = moving_target
    while index <= high:
        if array[index] < array[low]:
            array[index], array[moving_target] = array[moving_target], array[index]
            moving_target += 1
        index += 1
    moving_target -= 1
    array[low], array[moving_target] = array[moving_target], array[low]
    return moving_target


def sort(array, low, high):
    if low >= high:
        return array
    half = partition(array, low, high)
    sort(array, low, half - 1)
    sort(array, half + 1, high)
    return array


def quicksort(array):
    sort(array, 0, len(array) - 1)
    return array


def is_sorted(array):
    return array == sorted(array)


if __name__ == '__main__':
    size = 10000
    random_list = create_random_array(size)
    t0 = time.time()
    sorted_list = quicksort(random_list)
    t1 = time.time()
    if not is_sorted(sorted_list):
        raise Exception('The array was not correctly sorted.')
    if size < 20:
        print('Unsorted array   =>    {}'.format(random_list))
        print('Sorted array     =>    {}'.format(sorted_list))
    print('Took {:.3f}ms to sort {} elements.'.format((t1 - t0) * 1000, size))

File: javascript/sorting/quick-sort.js
"use strict";
const utils = require("../utils");

var testSizes = [100, 1000, 10000, 100000, 1000000];
var testTimeResults = [];

testSizes.forEach(n => {
  var toBeSorted, sorted, startTime, endTime;
  toBeSorted = utils.createRandomArray(n);
  startTime = +new Date();

  sorted = quickSort(toBeSorted);

  endTime = +new Date();
  console.log(`n=${n} took ${endTime - startTime}ms`);
  testTimeResults.push(endTime - startTime);

  // Check if the array is sorted
  utils.isSorted(sorted);
});

function swap(array, i, j) {
  var temp = array[i];
  array[i] = array[j];
  array[j] = temp;
}

/*
// This implementation appears on Algorithms Fourth Edition 
// by Robert Sedgewick and Kevin Wayne
function partition(array, low, high) {
  // array[low] is the pivot
  var lIndex = low;
  var rIndex = high + 1;
  while (true) {
    // From the left, search an element equal or bigger than the pivot
    while (array[++lIndex] <= array[low]) if (lIndex === high) break;
    // From the right, search an element smaller than the pivot
    while (array[--rIndex] > array[low]) if (rIndex === low) break;
    // This signal the end of the search
    if (lIndex >= rIndex) break;
    swap(array, lIndex, rIndex);
  }
  // Swap the pivot to its final position
  swap(array, rIndex, low);
  return rIndex;
}
*/

function partition(array, low, high) {
  // The pivot is the first element of the array (array[low])
  var searchIndex = low + 1;
  for (var i = searchIndex; i <= high; i++) {
    if (array[i] < array[low]) {
      swap(array, i, searchIndex);
      searchIndex++;
    }
  }
  searchIndex--;
  swap(array, low, searchIndex);
  return searchIndex;
}

function sort(array, low, high) {
  if (low >= high) return array;
  var partitionPivot = partition(array, low, high);
  sort(array, low, partitionPivot - 1);
  sort(array, partitionPivot + 1, high);
}

// Wrapper to avoid giving the user the responsability
// of computing the array boundaries
function quickSort(array) {
  sort(array, 0, array.length - 1);
  return array;
}
